crates/core/tedge_api/src/entity_store.rs

```diff
@@ -0,0 +1,539 @@
+            })
+            .map(|(k, _)| k.as_str())
+            .collect()
+    }
+
+    /// Updates entity store state based on the content of the entity
+    /// registration message.
+    ///
+    /// It can register a new entity in the store or update already registered
+    /// entity, returning a list of all entities affected by the update, e.g.:
+    ///
+    /// - when adding/removing a child device or service, the parent is affected
+    pub fn update(&mut self, message: EntityRegistrationMessage) -> Result<Vec<EntityId>, Error> {
+        if message.r#type == "device" && message.mqtt_id != self.main_device {
+            return Err(Error::MainDeviceAlreadyRegistered(
+                self.main_device.as_str().into(),
+            ));
+        }
+
+        let mut affected_entities = vec![];
+
+        let parent = match message.r#type.as_str() {
+            "child-device" => message.parent.or(Some(self.main_device.clone())),
+            "service" => message.parent.or(Some(self.main_device.clone())),
+            _ => None,
+        };
+
+        // parent device is affected if new device is its child
+        if let Some(parent) = &parent {
+            if !self.entities.contains_key(parent) {
+                return Err(Error::NoParent(parent.clone().into_boxed_str()));
+            }
+
+            affected_entities.push(parent.clone());
+        }
+
+        let external_id = self.external_id(&message.mqtt_id);
+        let entity_metadata = EntityMetadata {
+            r#type: message.r#type,
+            external_id: external_id.clone(),
+            parent,
+            other: message.payload,
+        };
+
+        // device is affected if it was previously registered and was updated
+        let previous = self
+            .entities
+            .insert(message.mqtt_id.clone(), entity_metadata);
+
+        if previous.is_some() {
+            affected_entities.push(message.mqtt_id);
+        } else {
+            self.external_id_index.insert(external_id, message.mqtt_id);
+        }
+
+        Ok(affected_entities)
+    }
+
+    /// An iterator over all registered entities.
+    pub fn iter(&self) -> impl Iterator<Item = (&EntityId, &EntityMetadata)> {
+        self.entities.iter()
+    }
+
+    /// Generate child device external ID.
+    ///
+    /// The external id is generated by prefixing the id with main device name
+    /// (device_common_name) and then appending the MQTT entity topic with `/`
+    ///  characters replaced by `:`.
+    ///
+    /// # Examples
+    /// - `device/main//` => `DEVICE_COMMON_NAME`
+    /// - `device/child001//` => `DEVICE_COMMON_NAME:device:child001`
+    /// - `device/child001/service/service001` => `DEVICE_COMMON_NAME:device:child001:service:service001`
+    /// - `factory01/hallA/packaging/belt001` => `DEVICE_COMMON_NAME:factory01:hallA:packaging:belt001`
+    fn external_id(&self, entity_topic: EntityIdRef) -> String {
```

Comment:
A better name for this function is `derive_external_id` as this is computation used when the registration message doesn't provide an id.